server/src/main/java/org/elasticsearch/script/Metadata.java

```diff
@@ -55,26 +55,50 @@ public class Metadata {
     public static FieldProperty<Number> LongField = new FieldProperty<>(Number.class).withValidation(FieldProperty.LONGABLE_NUMBER);
 
     protected final Map<String, Object> map;
-    protected final Map<String, FieldProperty<?>> properties;
+    private final Map<String, FieldProperty<?>> properties;
     protected static final FieldProperty<?> BAD_KEY = new FieldProperty<>(null, false, false, null);
 
-    public Metadata(Map<String, Object> map, Map<String, FieldProperty<?>> properties) {
+    /**
+     * Constructs a new Metadata object represented by the given map and properties.
+     * <p>
+     * The passed-in map is used directly -- subsequent modifications to it outside the methods of this class may result in
+     * undefined behavior.
+     * <p>
+     * The properties map is used directly as well, but we verify at runtime that it <b>must</b> be an immutable map (i.e. constructed
+     * via a call to {@link Map#of()} (or similar) in production, or via {@link Map#copyOf(Map)}} in tests). Since it must be an
+     * immutable map, subsequent modifications are not possible.
+     *
+     * @param map the backing map for this metadata instance
+     * @param properties the immutable map of defined properties for the type of metadata represented by this instance
+     */
+    protected Metadata(Map<String, Object> map, Map<String, FieldProperty<?>> properties) {
         this.map = map;
-        this.properties = Collections.unmodifiableMap(properties);
+        // we can't tell the compiler that properties must be a java.util.ImmutableCollections.AbstractImmutableMap, but
+        // we can use this copyOf + assert to verify that at runtime.
+        this.properties = Map.copyOf(properties);
+        assert this.properties == properties : "properties map must be constructed via Map.of(...) or Map.copyOf(...)";
```

Comment:
Can you add a test for this assert?

Reply:
Gah, yeah, I fall upon my sword. 0c97aa3